### PR TITLE
Add timestamp field to JUnitXML report

### PR DESCRIPTION
**Pull request by timcharper**
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/1.x/CONTRIBUTING.md) guidelines

Addresses #4153

### CODE REVIEW

File: testing/src/main/scala/sbt/JUnitXmlTestsListener.scala [hunk 1]
```diff
@@ -59,7 +61,7 @@ class JUnitXmlTestsListener(val outputDir: String) extends TestsListener {
    * Gathers data for one Test Suite. We map test groups to TestSuites.
    * Each TestSuite gets its own output file.
    */
-  class TestSuite(val name: String) {
+  class TestSuite(val name: String, timestamp: Long) {
```
Comment (eed3si9n) [inline testing/src/main/scala/sbt/JUnitXmlTestsListener.scala]: I'd suggest using `java.imte.LocalDateTime` here.

File: testing/src/main/scala/sbt/JUnitXmlTestsListener.scala [hunk 2]
```diff
@@ -211,3 +216,11 @@ class JUnitXmlTestsListener(val outputDir: String) extends TestsListener {
   /**Returns None*/
   override def contentLogger(test: TestDefinition): Option[ContentLogger] = None
 }
+
+object JUnitXmlTestsListener {
+  private val dateFormat = new SimpleDateFormat("yyyy-MM-dd")
+  private val timeFormat = new SimpleDateFormat("HH:mm:ss")
+
+  def formattedTimestamp(timestamp: Long): String =
+    s"${dateFormat.format(timestamp)}T${timeFormat.format(timestamp)}"
```
Comment (eed3si9n) [inline testing/src/main/scala/sbt/JUnitXmlTestsListener.scala]: Here's more JDK 8-ish time formatting code.


```scala
scala> import java.time.LocalDateTime
import java.time.LocalDateTime

scala> import java.time.format.DateTimeFormatter
import java.time.format.DateTimeFormatter

scala> import java.time.temporal.ChronoUnit
import java.time.temporal.ChronoUnit

scala> val t = LocalDateTime.now().truncatedTo(ChronoUnit.SECONDS)
t: java.time.LocalDateTime = 2018-05-12T01:39:18

scala> t.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
res0: String = 2018-05-12T01:39:18
```

**Review by eed3si9n — CHANGES_REQUESTED** [review]

File: testing/src/main/scala/sbt/JUnitXmlTestsListener.scala (hunk 1 above)
Comment (timcharper) [reply testing/src/main/scala/sbt/JUnitXmlTestsListener.scala]: Ahhh I avoided this because I didn’t know if sbt supports jdk 7 and older.

Comment (timcharper) [reply testing/src/main/scala/sbt/JUnitXmlTestsListener.scala]: Ahhh I avoided this because I didn’t know if sbt supports jdk 7 and older.

Comment (eed3si9n) [reply testing/src/main/scala/sbt/JUnitXmlTestsListener.scala]: sbt 1 is on Scala 2.12 so we can assume JDK 8 and up.

**Comment by timcharper** [comment]
@eed3si9n I've switched it to use Java time. Patch is much shorter.

File: testing/src/main/scala/sbt/JUnitXmlTestsListener.scala (hunk 2 above)
Comment (timcharper) [reply testing/src/main/scala/sbt/JUnitXmlTestsListener.scala]: Hmm, not sure if it's absolutely necessary to truncate to seconds. Wdyt?

Comment (timcharper) [reply testing/src/main/scala/sbt/JUnitXmlTestsListener.scala]: Oh, yep, it is. :) Good catch.

https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd#L15

File: testing/src/main/scala/sbt/JUnitXmlTestsListener.scala [hunk 3]
```diff
@@ -59,7 +60,7 @@ class JUnitXmlTestsListener(val outputDir: String) extends TestsListener {
    * Gathers data for one Test Suite. We map test groups to TestSuites.
    * Each TestSuite gets its own output file.
    */
-  class TestSuite(val name: String) {
+  class TestSuite(val name: String, timestamp: LocalDateTime) {
```
Comment (eed3si9n) [inline testing/src/main/scala/sbt/JUnitXmlTestsListener.scala:65]: Mima failed on Travis CI. It's unlikely that sbt plugins would instantiate this class, but adding another constructor that passes in `LocalDateTime.now()` should be easy to keep the bincompat, so could you plz add it?

**Review by eed3si9n — CHANGES_REQUESTED** [review]

**Comment by timcharper** [comment]
Sure, no problem

File: testing/src/main/scala/sbt/JUnitXmlTestsListener.scala (hunk 3 above)
Comment (timcharper) [reply testing/src/main/scala/sbt/JUnitXmlTestsListener.scala:65]: ✅

**Review by dwijnand — APPROVED** [review]
LGTM!

**Comment by dwijnand** [comment]
Leaving it to Eugene to :shipit:

**Comment by dwijnand** [comment]
I mean 🔀 (found a suitable "merge" emoji)

**Comment by timcharper** [comment]
Thanks @dwijnand !

I'm not sure why CI failed the build. 

https://travis-ci.org/sbt/sbt/jobs/378977478

```
:: retrieving :: org.scala-sbt#boot-scala
	confs: [default]
	5 artifacts copied, 0 already retrieved (18986kB/29ms)
Error: Could not retrieve sbt 1.1.4: missing sbt.xMain
restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
The command "sbt -J-XX:ReservedCodeCacheSize=128m -J-Xmx800M -J-Xms800M -J-server "$SBT_CMD"" exited with 1.
```

🤔

**Review by eed3si9n — APPROVED** [review]
LGTM pending Travis CI

**Comment by eed3si9n** [comment]
Something looks odd with Travis.

```
Getting org.scala-sbt sbt 1.1.4  (this may take some time)...
:: retrieving :: org.scala-sbt#boot-app
	confs: [default]
	0 artifacts copied, 76 already retrieved (0kB/67ms)
Error: Could not retrieve sbt 1.1.4: missing sbt.xMain
```

I am guessing this is unrelated to your change.

**Comment by eed3si9n** [comment]
Wiping Travis CI cache seemed to have solved it, so it was corrupt Ivy cache, I am guessing.

**Comment by eed3si9n** [comment]
Thanks @timcharper for the contribution.
Check out https://github.com/sbt/sbt/issues for more ideas :)

**Pull request merged**